src/fastalloc/mod.rs

```diff
@@ -0,0 +1,1793 @@
+use core::convert::TryInto;
+use core::iter::FromIterator;
+use core::ops::{Index, IndexMut};
+use crate::domtree::dominates;
+use crate::{domtree, postorder, AllocationKind, Block, Inst, InstPosition, Operand, OperandConstraint, OperandKind, OperandPos, PReg, PRegSet, RegClass, SpillSlot, VReg};
+use crate::{Function, MachineEnv, ssa::validate_ssa, ProgPoint, Edit, Output};
+use crate::{cfg::CFGInfo, RegAllocError, Allocation, ion::Stats};
+use alloc::collections::{BTreeSet, VecDeque};
+use alloc::vec::Vec;
+use hashbrown::{HashSet, HashMap};
+
+use std::println;
+use std::format;
+
+mod lru;
+mod iter;
+use lru::*;
+use iter::*;
+
+#[derive(Debug)]
+struct Allocs {
+    allocs: Vec<Allocation>,
+    /// `inst_alloc_offsets[i]` is the offset into `allocs` for the allocations of
+    /// instruction `i`'s operands.
+    inst_alloc_offsets: Vec<u32>,
+}
+
+impl Allocs {
+    fn new<F: Function>(func: &F, env: &MachineEnv) -> Self {
+        // The number of operands is <= number of virtual registers
```

Comment:
I don't think this is actually true -- we could have more operands than virtual registers because of uses (imagine we have a function with a single vreg `v1`, and a instruction that uses `v1` a million times). It shouldn't really matter here since this is just a pre-allocation heuristic; but it may be better to do something like `func.num_insts() * 3` as a first guess.

Reply:
Ah, yes. This is an outdated comment.